contracts/external/cw-payroll/src/contract.rs

```diff
@@ -161,71 +277,79 @@ pub fn execute_receive(
         } => execute_create_stream(
             env,
             deps,
-            config,
-            // TODO: Change this, but still conform to standard
             StreamParams {
-                admin: wrapped.sender,
+                admin: admin.unwrap_or(wrapped.sender.clone()),
                 recipient,
-                balance: wrapped.amount,
+                balance: WrappedBalance::from(wrapped),
                 start_time,
                 end_time,
                 title: None,
-                description: None
+                description: None,
             },
         ),
     }
 }
 
-pub fn execute_distribute(
-    env: Env,
-    deps: DepsMut,
-    info: MessageInfo,
-    id: u64,
-) -> Result<Response, ContractError> {
+pub fn execute_distribute(env: Env, deps: DepsMut, id: u64) -> Result<Response, ContractError> {
     let mut stream = STREAMS
         .may_load(deps.storage, id)?
         .ok_or(ContractError::StreamNotFound {})?;
 
-    // TODO: Why is this required? is there a security issue here if someone pays to send you tokens all the time?
-    // NOTE: commenting out for simple testing with croncat, if required will add approved list.
-    // if stream.recipient != info.sender {
-    //     return Err(ContractError::NotStreamRecipient {
-    //         recipient: stream.recipient,
-    //     });
-    // }
-
-    // TODO: Def change to better comparitor
-    if stream.claimed_balance >= stream.balance {
-        return Err(ContractError::StreamFullyClaimed {});
+    let mut msgs: Vec<CosmosMsg> = vec![];
+    let (available_claims, _) = stream.calc_distribution_rate(env.block.time);
+
+    if !stream.can_ditribute_more() || available_claims.u128() <= 0 {
+        return Err(ContractError::NoFundsToClaim {
+            claimed: stream.claimed_balance,
+        });
     }
+    if let Some(native_balance) = stream.balance.native() {
+        let to_claim = Coin {
+            denom: native_balance.denom.clone(),
+            amount: available_claims,
+        };
 
-    let block_time = env.block.time.seconds();
-    let time_passed = std::cmp::min(block_time, stream.end_time).saturating_sub(stream.start_time);
-    let vested = Uint128::from(time_passed) * stream.rate_per_second;
-    let released = vested - stream.claimed_balance;
+        stream
+            .claimed_balance
+            .checked_add_native(&[to_claim.clone()])
+            .unwrap();
+        let msg = CosmosMsg::Bank(BankMsg::Send {
+            to_address: stream.recipient.clone().into(),
+            amount: vec![Coin {
+                denom: native_balance.denom.clone(),
+                amount: to_claim.amount,
```

Comment:
I believe this value to be wrong @deveusss , but please double check my thinking.

You are appropriately checking above for claimed balance & current stream balance, so dont change those.

On this line, `to_claim.amount` doesnt take into account the amount already claimed. So if i hit the `distribute` method multiple times, i can immediately drain the full stream i think.
Line 299 with `available_claims` needs to subtract `stream.claimed_balance` to be fully accurate on line 309.

If im wrong, please help me understand :)
Would also love test coverage for this re-distribute multiple times in a row case.

Reply:
Also, noting this is the case for cw20 impl on line 328